cors: make enforced allow origins optional

erquhart opened the pull request:
The cors helper returns a 403 from requests that don't match allowedOrigins. This can cause issues for server-to-server requests, and goes beyond basic cors functionality.

This PR adds an `enforceAllowOrigins` option that defaults to true to avoid breaking.

ianmacartney commented:
nit: I'd rather have options that default to false, so not setting it & setting it to false are the same and !!opts.optionalOption works.

I did a quick survey and I think the behavior we can go straight to is:
If there is no origin header set, skip CORS entirely (don't return any headers).
I'm guessing that in these scenarios they're not even calling OPTIONS first anyways

ianmacartney commented on packages/convex-helpers/server/cors.ts:
```diff
@@ -330,7 +336,7 @@ const handleCors = ({
         }
       }
 
```
```suggestion
const innerHandler = ("_handler" in originalHandler
        ? (originalHandler["_handler"] as PublicHttpAction)
        : originalHandler) as unknown as (
        ctx: GenericActionCtx<any>,
        request: Request,
      ) => Promise<Response>;
      const originalResponse = await innerHandler(ctx, request);
      if (!requestOrigin) {
        return originalResponse;
      }
```
but put it above the origing matching

erquhart commented:
@ianmacartney agree on default to false being better, I only did it this way to avoid a breaking change. If we want to cold turkey it I can change the default.

Handling no origin differently and not returning headers works, I just don't know if it has an impact since a server will generally ignore the cors headers either way. There may be edge cases I don't know about here, though.

ianmacartney commented:
I trust tom’s take on this and happy to make the breaking change.
Thankfully no requests will start to fail. It is only breaking for people
who are actively rejecting this sort of traffic, which I can’t think of a
use case for

On Thu, Jun 19, 2025 at 12:09 AM Shawn Erquhart ***@***.***>
wrote:

> *erquhart* left a comment (get-convex/convex-helpers#657)
> <https://github.com/get-convex/convex-helpers/pull/657#issuecomment-2986909311>
>
> @ianmacartney <https://github.com/ianmacartney> agree on default to false
> being better, I only did it this way to avoid a breaking change. If we want
> to cold turkey it I can change the default.
>
> Handling no origin differently and not returning headers works, I just
> don't know if it has an impact since a server will generally ignore the
> cors headers either way. There may be edge cases I don't know about here,
> though.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/get-convex/convex-helpers/pull/657#issuecomment-2986909311>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACZQW3NVHL5S4ISZELWVPL3EJO37AVCNFSM6AAAAAB7UAO2WCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSOBWHEYDSMZRGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ianmacartney commented:
Honestly when I first made it I probably should’ve looked at a few
reference implementations. It might still be worth asking AI to compare
this implementation to Hono’s or other as a fast sanity check and ask about
missing origin. I’m both convinced that dropping headers with no origin is
fine and have a general fear of security incidents

On Thu, Jun 19, 2025 at 9:11 AM Shawn Erquhart ***@***.***>
wrote:

> *erquhart* left a comment (get-convex/convex-helpers#657)
> <https://github.com/get-convex/convex-helpers/pull/657#issuecomment-2988607980>
>
> @ianmacartney <https://github.com/ianmacartney> I originally looked
> through this on my phone and completely missed that you provided an
> implementation to go with. I'll switch the default to false and use the
> implementation provided.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/get-convex/convex-helpers/pull/657#issuecomment-2988607980>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACZQW3QXPSE2ZFN6XOXEA33ELOJJAVCNFSM6AAAAAB7UAO2WCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSOBYGYYDOOJYGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

erquhart commented:
I spent a good amount of time on cors in the last week troubleshooting better auth issues, and one standout principle I came away with is that anything that can send a cross origin request with no origin header can also manipulate the origin header anyway. So cors adds no security whatsoever for any client other than a secure modern browser.

I was really confused with Hono's behavior when I was working on the better auth / convex / hono doc until I understood this, because it just worked for all server to server requests. Theirs is actually simpler, I kinda want to copy it honestly: https://github.com/honojs/hono/blob/main/src/middleware/cors/index.ts

They set the allow origin header on responses dynamically, only if there's a match (origin header or not). The others cors headers are still added regardless of origin match.

ianmacartney commented:
Keeping the other headers or not returning any both sound safe - up to you
which direction you want to go

On Thu, Jun 19, 2025 at 4:01 PM Shawn Erquhart ***@***.***>
wrote:

> *erquhart* left a comment (get-convex/convex-helpers#657)
> <https://github.com/get-convex/convex-helpers/pull/657#issuecomment-2989323311>
>
> I spent a good amount of time on cors in the last week troubleshooting
> better auth issues, and one standout principle I came away with is that
> anything that can send a cross origin request with no origin header can
> also manipulate the origin header anyway. So cors adds no security
> whatsoever for any client other than a secure modern browser.
>
> I was really confused with Hono's behavior when I was working on the
> better auth / convex / hono doc until I understood this, because it just
> worked for all server to server requests. Theirs is actually simpler, I
> kinda want to copy it honestly:
> https://github.com/honojs/hono/blob/main/src/middleware/cors/index.ts
>
> They set the allow origin header on responses dynamically, only if there's
> a match (origin header or not). The others cors headers are still added
> regardless of origin match.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/get-convex/convex-helpers/pull/657#issuecomment-2989323311>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACZQW6P4DZ5S6JZ7I3QEN33EM6NBAVCNFSM6AAAAAB7UAO2WCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSOBZGMZDGMZRGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

erquhart commented:
I reviewed this with AI assistance compared to Hono and came away with three differences worth addressing:

1. Hono sets the allow origins response header in a non-disclosing way - for allowed origins, it's set to the request origin, otherwise the header is omitted. Namely, "*" is never returned.
2. The vary header is handled differently, so I just pulled in Hono's handling for that header:
> More nuanced. It only sets Vary: Origin if the origin option is not the wildcard '*'. It also correctly appends to an existing Vary header rather than overwriting it. Furthermore, if it reflects request headers, it also adds Vary: Access-Control-Request-Headers.
3. Hono accepts a function for setting the allow origin header. I had already implemented this in the copy I'm running for the Better Auth integration to support Better Auth's similar handling of their `trustedOrigins` option, which we use to set defaults for allow origins. With this in place I can just run the helper directly.

I've added these changes and associated test updates. Let me know if you'd rather not do all this and I can slim the PR back down and keep running a copy in the BA component, that's not a problem.

ianmacartney commented on packages/convex-helpers/server/cors.ts:
```diff
@@ -372,18 +401,25 @@ const handleCors = ({
         newHeaders.set(key, value);
       });
 
+      if (debug) {
+        console.log("CORS response headers", newHeaders);
+      }
+
```
```suggestion
```

ianmacartney reviewed (approved):
All sounds good to me - and you think folks will still want `enforceAllowOrigins`?

ianmacartney commented:
It's out!
And I added a CLA template explicitly now - we'll see, but I don't think
it's necessary for folks we have existing agreements with

On Sat, Jun 21, 2025 at 11:30 PM Shawn Erquhart ***@***.***>
wrote:

> *erquhart* left a comment (get-convex/convex-helpers#657)
> <https://github.com/get-convex/convex-helpers/pull/657#issuecomment-2993970215>
>
> @ianmacartney <https://github.com/ianmacartney> I did consider dropping
> enforcement altogether, I can open another PR to do that.
>
> Don't know what's up with the cla text thing, I'll just include it next
> time.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/get-convex/convex-helpers/pull/657#issuecomment-2993970215>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACZQW64D6C63BJCQ4IOWMD3EZEO7AVCNFSM6AAAAAB7UAO2WCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSOJTHE3TAMRRGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
